File: backend/src/main/java/de/interaapps/pastefy/model/database/BackgroundJob.java
package de.interaapps.pastefy.model.database;

import org.javawebstack.orm.Model;
import org.javawebstack.orm.annotation.Column;
import org.javawebstack.orm.annotation.Dates;
import org.javawebstack.orm.annotation.Table;

import java.sql.Timestamp;

@Dates
@Table("background_jobs")
public class BackgroundJob extends Model {
    @Column(id = true, ai = false, size = 255)
    public String key;

    @Column
    public Type type;

    @Column
    public int entityId;

    @Column
    public int sourceVersion;

    @Column
    public int promptVersion;

    @Column
    public Status status = Status.PENDING;

    @Column
    public int attempts = 0;

    @Column
    public Timestamp availableAt;

    @Column
    public Timestamp leaseUntil;

    @Column(size = 64)
    public String leaseToken;

    @Column(size = 2048)
    public String lastError;

    @Column
    public Timestamp createdAt;

    @Column
    public Timestamp updatedAt;

    public enum Type {
        PASTE_AI_INFO
    }

    public enum Status {
        PENDING,
        RUNNING,
        DONE,
        FAILED
    }
}

File: backend-kt/src/main/kotlin/de/interaapps/pastefy/entities/BackgroundJob.kt
package de.interaapps.pastefy.entities

import jakarta.persistence.*
import java.time.Instant

@Entity
@Table(
    name = "pastefy_background_jobs",
    indexes = [
        Index(name = "pastefy_background_jobs_status_available_at_index", columnList = "status, available_at"),
        Index(name = "pastefy_background_jobs_status_lease_until_index", columnList = "status, lease_until"),
        Index(name = "pastefy_background_jobs_type_entity_id_index", columnList = "type, entity_id"),
    ],
)
class BackgroundJob(
    @Id @Column(name = "`key`", length = 255) var key: String = "",
    @Enumerated(EnumType.STRING) @Column(nullable = false) var type: Type = Type.PASTE_AI_INFO,
    @Column(nullable = false) var entityId: Int = 0,
    @Column(nullable = false) var sourceVersion: Int = 0,
    @Column(nullable = false) var promptVersion: Int = 0,
    @Enumerated(EnumType.STRING) @Column(nullable = false) var status: Status = Status.PENDING,
    @Column(nullable = false) var attempts: Int = 0,
    var availableAt: Instant? = null,
    var leaseUntil: Instant? = null,
    @Column(length = 64) var leaseToken: String? = null,
    @Column(length = 2048) var lastError: String? = null,
    @Column(nullable = false, updatable = false) var createdAt: Instant? = null,
    @Column(nullable = false) var updatedAt: Instant? = null,
) {
    @PrePersist fun prePersist() {
        val now = Instant.now()
        if (createdAt == null) createdAt = now
        updatedAt = now
    }
    @PreUpdate fun preUpdate() { updatedAt = Instant.now() }

    enum class Type { PASTE_AI_INFO }
    enum class Status { PENDING, RUNNING, DONE, FAILED }
}
